app/src/main/java/com/android/sample/model/user/UserRepositoryFirestore.kt

```diff
@@ -0,0 +1,86 @@
+package com.android.sample.model.user
+
+import android.util.Log
+import com.google.android.gms.tasks.Task
+import com.google.firebase.Firebase
+import com.google.firebase.auth.auth
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+
+class UserRepositoryFirestore(private val db: FirebaseFirestore) : UserRepository {
+  private val collectionPath = "users"
+
+  override fun init(onSuccess: () -> Unit) {
+    if (Firebase.auth.currentUser != null) {
+      onSuccess()
+    }
+  }
+
+  override fun getUserById(id: String, onSuccess: (User) -> Unit, onFailure: (Exception) -> Unit) {
+    val doc = db.collection(collectionPath).document(id)
+    doc.get().addOnCompleteListener { task ->
+      if (task.isSuccessful) {
+        onSuccess(convertSnapshot(id, task.result))
+      } else {
+        task.exception?.let { onFailure(it) }
+      }
+    }
+  }
+
+  override fun addUser(user: User, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    performUserDatabaseAction(
+        db.collection(collectionPath).document(user.uid).set(user), onSuccess, onFailure)
+  }
+
+  override fun updateUser(user: User, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    performUserDatabaseAction(
+        db.collection(collectionPath).document(user.uid).set(user), onSuccess, onFailure)
```

Comment:
Because we update periodically, we set all the data at once every period. To make the server's workload lighter, I made it merge the differences instead of completely replace at each update.